File: Sources/HoverPocket/Capabilities/CapabilityBrokerPersistence.swift
import Foundation

enum CapabilityLedgerStart<Result> {
    case execute
    case replay(Result)
    case unknown
}

final class CapabilityBrokerLedger {
    private enum RecordState: String, Codable {
        case pending
        case completed
    }

    private struct InvocationRecord: Codable {
        let planDigest: String
        let argumentDigest: String
        let capability: PocketCapabilityKey
        var state: RecordState
        var receipt: CapabilityReceipt?
        var completedAt: Date?
    }

    private struct WorkflowRecord: Codable {
        let planDigest: String
        var state: RecordState
        var receipt: CapabilityWorkflowReceipt?
        var completedAt: Date?
    }

    private struct State: Codable {
        var version = 2
        var invocations: [String: InvocationRecord] = [:]
        var workflows: [String: WorkflowRecord] = [:]
    }

    private let fileURL: URL
    private var state: State
    private let encoder: JSONEncoder
    private let decoder: JSONDecoder
    private let lock = NSLock()

    init(rootDirectory: URL) throws {
        self.fileURL = rootDirectory.appendingPathComponent("capability-broker-ledger.json", isDirectory: false)
        self.encoder = JSONEncoder()
        self.decoder = JSONDecoder()
        encoder.dateEncodingStrategy = .iso8601
        encoder.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]
        decoder.dateDecodingStrategy = .iso8601

        do {
            try FileManager.default.createDirectory(
                at: rootDirectory,
                withIntermediateDirectories: true,
                attributes: [.posixPermissions: 0o700]
            )
            if FileManager.default.fileExists(atPath: fileURL.path) {
                let data = try Data(contentsOf: fileURL)
                state = try decoder.decode(State.self, from: data)
                guard state.version == 1 || state.version == 2 else {
                    throw CapabilityBrokerError.ledgerUnavailable
                }
                if state.version == 1 {
                    for key in state.invocations.keys {
                        guard var record = state.invocations[key] else { continue }
                        record.completedAt = record.receipt?.completedAt
                        state.invocations[key] = record
                    }
                    for key in state.workflows.keys {
                        guard var record = state.workflows[key] else { continue }
                        record.completedAt = record.receipt?.completedAt
                        state.workflows[key] = record
                    }
                    state.version = 2
                    try persistUnlocked()
                }
            } else {
                state = State()
            }
        } catch let error as CapabilityBrokerError {
            throw error
        } catch {
            throw CapabilityBrokerError.ledgerUnavailable
        }
    }

    func beginInvocation(
        idempotencyKey: String,
        planDigest: String,
        argumentDigest: String,
        capability: PocketCapabilityKey
    ) throws -> CapabilityLedgerStart<CapabilityReceipt> {
        try synchronized {
            if let existing = state.invocations[idempotencyKey] {
                guard existing.planDigest == planDigest,
                      existing.argumentDigest == argumentDigest,
                      existing.capability == capability else {
                    throw CapabilityBrokerError.idempotencyConflict(idempotencyKey)
                }
                if existing.state == .completed, let receipt = existing.receipt {
                    return .replay(receipt.replayCopy())
                }
                return .unknown
            }

            state.invocations[idempotencyKey] = InvocationRecord(
                planDigest: planDigest,
                argumentDigest: argumentDigest,
                capability: capability,
                state: .pending,
                receipt: nil,
                completedAt: nil
            )
            try persistUnlocked()
            return .execute
        }
    }

    func completeInvocation(idempotencyKey: String, receipt: CapabilityReceipt) throws {
        try synchronized {
            guard var record = state.invocations[idempotencyKey],
                  record.planDigest == receipt.planDigest,
                  record.capability == receipt.capability else {
                throw CapabilityBrokerError.ledgerUnavailable
            }
            record.state = .completed
            record.receipt = receipt.durableCopy()
            record.completedAt = receipt.completedAt
            state.invocations[idempotencyKey] = record
            try persistUnlocked()
        }
    }

    func lookupWorkflow(planID: String, planDigest: String) throws -> CapabilityLedgerStart<CapabilityWorkflowReceipt> {
        try synchronized {
            if let existing = state.workflows[planID] {
                guard existing.planDigest == planDigest else {
                    throw CapabilityBrokerError.idempotencyConflict(planID)
                }
                if existing.state == .completed, let receipt = existing.receipt {
                    return .replay(receipt.replayCopy())
                }
                return .unknown
            }
            return .execute
        }
    }

    func startWorkflow(planID: String, planDigest: String) throws {
        try synchronized {
            guard state.workflows[planID] == nil else {
                throw CapabilityBrokerError.idempotencyConflict(planID)
            }
            state.workflows[planID] = WorkflowRecord(
                planDigest: planDigest,
                state: .pending,
                receipt: nil,
                completedAt: nil
            )
            try persistUnlocked()
        }
    }

    func completeWorkflow(_ receipt: CapabilityWorkflowReceipt) throws {
        try synchronized {
            guard var record = state.workflows[receipt.planID],
                  record.planDigest == receipt.planDigest else {
                throw CapabilityBrokerError.ledgerUnavailable
            }
            record.state = .completed
            record.receipt = receipt.durableCopy()
            record.completedAt = receipt.completedAt
            state.workflows[receipt.planID] = record
            try persistUnlocked()
        }
    }

    @discardableResult
    func redactCompletedReceipts(
        olderThan cutoff: Date?,
        redactAll: Bool = false
    ) throws -> Int {
        try synchronized {
            var redacted = 0
            for key in state.invocations.keys {
                guard var record = state.invocations[key],
                      record.state == .completed,
                      record.receipt != nil,
                      redactAll || cutoff.map({ (record.completedAt ?? .distantPast) < $0 }) == true else {
                    continue
                }
                record.receipt = nil
                record.completedAt = nil
                state.invocations[key] = record
                redacted += 1
            }
            for key in state.workflows.keys {
                guard var record = state.workflows[key],
                      record.state == .completed,
                      record.receipt != nil,
                      redactAll || cutoff.map({ (record.completedAt ?? .distantPast) < $0 }) == true else {
                    continue
                }
                record.receipt = nil
                record.completedAt = nil
                state.workflows[key] = record
                redacted += 1
            }
            if redacted > 0 {
                try persistUnlocked()
            }
            return redacted
        }
    }

    func retentionSnapshot() -> CapabilityLedgerRetentionSnapshot {
        synchronized {
            let invocationRecords = Array(state.invocations.values)
            let workflowRecords = Array(state.workflows.values)
            let completedRecords = invocationRecords.map { ($0.state, $0.receipt != nil) }
                + workflowRecords.map { ($0.state, $0.receipt != nil) }
            return CapabilityLedgerRetentionSnapshot(
                storedReceiptCount: completedRecords.filter { $0.0 == .completed && $0.1 }.count,
                redactedTombstoneCount: completedRecords.filter { $0.0 == .completed && !$0.1 }.count,
                pendingCount: completedRecords.filter { $0.0 == .pending }.count
            )
        }
    }

    private func persistUnlocked() throws {
        do {
            let data = try encoder.encode(state)
            try data.write(to: fileURL, options: .atomic)
            try FileManager.default.setAttributes([.posixPermissions: 0o600], ofItemAtPath: fileURL.path)
        } catch {
            throw CapabilityBrokerError.ledgerUnavailable
        }
    }

    private func synchronized<T>(_ body: () throws -> T) rethrows -> T {
        lock.lock()
        defer { lock.unlock() }
        return try body()
    }
}

final class CapabilityApprovalStore {
    enum Decision {
        case approve
        case reject
    }

    private struct PendingApproval {
        let request: CapabilityApprovalRequest
    }

    private struct IssuedGrant {
        let requestID: String
        let planID: String
        let planDigest: String
        let principal: CapabilityPrincipal
        let appContext: CapabilityAppContext?
        let permissions: Set<String>
        let expiresAt: Date
    }

    private var pending: [String: PendingApproval] = [:]
    private var grants: [String: IssuedGrant] = [:]
    private var consumedTokens: Set<String> = []
    private let timeToLive: TimeInterval

    init(timeToLive: TimeInterval = 300) {
        self.timeToLive = timeToLive
    }

    func pendingRequest(requestID: String) -> CapabilityApprovalRequest? {
        pending[requestID]?.request
    }

    func discardPending(requestID: String) {
        pending.removeValue(forKey: requestID)
    }

    func request(
        for plan: CapabilityExecutionPlan,
        digest: String,
        descriptors: [PocketCapabilityDescriptor],
        now: Date
    ) throws -> CapabilityApprovalRequest? {
        let effects = try zip(plan.steps, descriptors).compactMap { step, descriptor -> CapabilityApprovalEffect? in
            guard descriptor.approvalPolicy.requiresExecutionApproval else { return nil }
            return CapabilityApprovalEffect(
                stepID: step.id,
                capability: step.capability,
                effect: descriptor.effect,
                argumentDigest: try CapabilityCanonicalJSON.digest(step.arguments),
                summaryKey: "approval.\(step.capability.id)",
                rollbackAvailable: descriptor.rollbackAvailable
            )
        }
        guard !effects.isEmpty else { return nil }

        let request = CapabilityApprovalRequest(
            id: "approval:\(UUID().uuidString.lowercased())",
            planID: plan.id,
            planDigest: digest,
            principal: plan.principal,
            appContext: plan.appContext,
            createdAt: now,
            expiresAt: now.addingTimeInterval(timeToLive),
            nonce: "nonce:\(UUID().uuidString.lowercased())",
            effects: effects,
            requiredPermissions: Set(zip(descriptors, plan.steps).flatMap { descriptor, _ in
                descriptor.approvalPolicy.requiresExecutionApproval ? descriptor.permissions : []
            })
        )
        pending[request.id] = PendingApproval(request: request)
        return request
    }

    func decide(
        requestID: String,
        presentedPlanDigest: String,
        decision: Decision,
        now: Date
    ) throws -> CapabilityApprovalGrant {
        guard let pendingApproval = pending.removeValue(forKey: requestID) else {
            throw CapabilityBrokerError.approvalInvalid
        }
        let request = pendingApproval.request
        guard request.planDigest == presentedPlanDigest else {
            throw CapabilityBrokerError.approvalInvalid
        }
        guard now <= request.expiresAt else {
            throw CapabilityBrokerError.approvalExpired
        }
        guard decision == .approve else {
            throw CapabilityBrokerError.approvalRejected
        }

        let token = "grant:\(UUID().uuidString.lowercased())"
        grants[token] = IssuedGrant(
            requestID: request.id,
            planID: request.planID,
            planDigest: request.planDigest,
            principal: request.principal,
            appContext: request.appContext,
            permissions: request.requiredPermissions,
            expiresAt: request.expiresAt
        )
        return CapabilityApprovalGrant(token: token)
    }

    func consume(
        _ grant: CapabilityApprovalGrant,
        plan: CapabilityExecutionPlan,
        digest: String,
        requiredPermissions: Set<String>,
        now: Date
    ) throws {
        if consumedTokens.contains(grant.token) {
            throw CapabilityBrokerError.approvalReplayed
        }
        guard let issued = grants.removeValue(forKey: grant.token),
              issued.planID == plan.id,
              issued.planDigest == digest,
              issued.principal == plan.principal,
              issued.appContext == plan.appContext,
              requiredPermissions.isSubset(of: issued.permissions) else {
            throw CapabilityBrokerError.approvalInvalid
        }
        consumedTokens.insert(grant.token)
        if consumedTokens.count > 1_024 {
            consumedTokens.removeAll(keepingCapacity: true)
            consumedTokens.insert(grant.token)
        }
        guard now <= issued.expiresAt else {
            throw CapabilityBrokerError.approvalExpired
        }
    }
}

File: windows/src/HoverPocket.Shell/Capabilities/CapabilityBrokerPersistence.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.RegularExpressions;

namespace HoverPocket.Shell.Capabilities;

internal enum CapabilityLedgerStartKind
{
    Execute,
    Replay,
    Unknown
}

internal sealed record CapabilityLedgerStart<T>(CapabilityLedgerStartKind Kind, T? Receipt = default);

internal sealed class CapabilityBrokerLedger
{
    private enum RecordState
    {
        Pending,
        Completed
    }

    private sealed record InvocationRecord(
        string PlanDigest,
        string ArgumentDigest,
        PocketCapabilityKey Capability,
        RecordState State,
        CapabilityReceipt? Receipt,
        DateTimeOffset? CompletedAt);

    private sealed record WorkflowRecord(
        string PlanDigest,
        RecordState State,
        CapabilityWorkflowReceipt? Receipt,
        DateTimeOffset? CompletedAt);

    private sealed class LedgerState
    {
        public int Version { get; set; } = 2;
        public Dictionary<string, InvocationRecord> Invocations { get; set; } = new(StringComparer.Ordinal);
        public Dictionary<string, WorkflowRecord> Workflows { get; set; } = new(StringComparer.Ordinal);
    }

    private static readonly JsonSerializerOptions Options = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        WriteIndented = false
    };

    private readonly string _filePath;
    private readonly object _sync = new();

    private LedgerState _state;

    public CapabilityBrokerLedger(string rootDirectory)
    {
        try
        {
            Directory.CreateDirectory(rootDirectory);
            _filePath = Path.Combine(rootDirectory, "capability-broker-ledger.json");
            _state = File.Exists(_filePath)
                ? JsonSerializer.Deserialize<LedgerState>(File.ReadAllBytes(_filePath), Options)
                    ?? throw Unavailable()
                : new LedgerState();
            if (_state.Version is not 1 and not 2)
            {
                throw Unavailable();
            }
            if (_state.Version == 1)
            {
                foreach (var key in _state.Invocations.Keys.ToArray())
                {
                    var record = _state.Invocations[key];
                    _state.Invocations[key] = record with { CompletedAt = record.Receipt?.CompletedAt };
                }
                foreach (var key in _state.Workflows.Keys.ToArray())
                {
                    var record = _state.Workflows[key];
                    _state.Workflows[key] = record with { CompletedAt = record.Receipt?.CompletedAt };
                }
                _state.Version = 2;
                Persist();
            }
        }
        catch (CapabilityBrokerException)
        {
            throw;
        }
        catch
        {
            throw Unavailable();
        }
    }

    public CapabilityLedgerStart<CapabilityReceipt> BeginInvocation(
        string idempotencyKey,
        string planDigest,
        string argumentDigest,
        PocketCapabilityKey capability)
    {
        lock (_sync)
        {
            if (_state.Invocations.TryGetValue(idempotencyKey, out var existing))
            {
                if (existing.PlanDigest != planDigest
                    || existing.ArgumentDigest != argumentDigest
                    || existing.Capability != capability)
                {
                    throw new CapabilityBrokerException("CAPABILITY_IDEMPOTENCY_CONFLICT", idempotencyKey);
                }
                return existing.State == RecordState.Completed && existing.Receipt is not null
                    ? new CapabilityLedgerStart<CapabilityReceipt>(CapabilityLedgerStartKind.Replay, existing.Receipt.ReplayCopy())
                    : new CapabilityLedgerStart<CapabilityReceipt>(CapabilityLedgerStartKind.Unknown);
            }
            _state.Invocations[idempotencyKey] = new InvocationRecord(
                planDigest,
                argumentDigest,
                capability,
                RecordState.Pending,
                null,
                null);
            Persist();
            return new CapabilityLedgerStart<CapabilityReceipt>(CapabilityLedgerStartKind.Execute);
        }
    }

    public void CompleteInvocation(string idempotencyKey, CapabilityReceipt receipt)
    {
        lock (_sync)
        {
            if (!_state.Invocations.TryGetValue(idempotencyKey, out var existing)
                || existing.PlanDigest != receipt.PlanDigest
                || existing.Capability != receipt.Capability)
            {
                throw Unavailable();
            }
            _state.Invocations[idempotencyKey] = existing with
            {
                State = RecordState.Completed,
                Receipt = receipt.DurableCopy(),
                CompletedAt = receipt.CompletedAt
            };
            Persist();
        }
    }

    public CapabilityLedgerStart<CapabilityWorkflowReceipt> LookupWorkflow(string planId, string planDigest)
    {
        lock (_sync)
        {
            if (!_state.Workflows.TryGetValue(planId, out var existing))
            {
                return new CapabilityLedgerStart<CapabilityWorkflowReceipt>(CapabilityLedgerStartKind.Execute);
            }
            if (existing.PlanDigest != planDigest)
            {
                throw new CapabilityBrokerException("CAPABILITY_IDEMPOTENCY_CONFLICT", planId);
            }
            return existing.State == RecordState.Completed && existing.Receipt is not null
                ? new CapabilityLedgerStart<CapabilityWorkflowReceipt>(CapabilityLedgerStartKind.Replay, existing.Receipt.ReplayCopy())
                : new CapabilityLedgerStart<CapabilityWorkflowReceipt>(CapabilityLedgerStartKind.Unknown);
        }
    }

    public void StartWorkflow(string planId, string planDigest)
    {
        lock (_sync)
        {
            if (!_state.Workflows.TryAdd(planId, new WorkflowRecord(planDigest, RecordState.Pending, null, null)))
            {
                throw new CapabilityBrokerException("CAPABILITY_IDEMPOTENCY_CONFLICT", planId);
            }
            Persist();
        }
    }

    public void CompleteWorkflow(CapabilityWorkflowReceipt receipt)
    {
        lock (_sync)
        {
            if (!_state.Workflows.TryGetValue(receipt.PlanId, out var existing)
                || existing.PlanDigest != receipt.PlanDigest)
            {
                throw Unavailable();
            }
            _state.Workflows[receipt.PlanId] = existing with
            {
                State = RecordState.Completed,
                Receipt = receipt.DurableCopy(),
                CompletedAt = receipt.CompletedAt
            };
            Persist();
        }
    }

    public int RedactCompletedReceipts(DateTimeOffset? olderThan, bool redactAll = false)
    {
        lock (_sync)
        {
            var redacted = 0;
            foreach (var key in _state.Invocations.Keys.ToArray())
            {
                var record = _state.Invocations[key];
                if (record.State != RecordState.Completed
                    || record.Receipt is null
                    || (!redactAll && (olderThan is null || (record.CompletedAt ?? DateTimeOffset.MinValue) >= olderThan.Value)))
                {
                    continue;
                }
                _state.Invocations[key] = record with { Receipt = null, CompletedAt = null };
                redacted += 1;
            }
            foreach (var key in _state.Workflows.Keys.ToArray())
            {
                var record = _state.Workflows[key];
                if (record.State != RecordState.Completed
                    || record.Receipt is null
                    || (!redactAll && (olderThan is null || (record.CompletedAt ?? DateTimeOffset.MinValue) >= olderThan.Value)))
                {
                    continue;
                }
                _state.Workflows[key] = record with { Receipt = null, CompletedAt = null };
                redacted += 1;
            }
            if (redacted > 0)
            {
                Persist();
            }
            return redacted;
        }
    }

    public CapabilityLedgerRetentionSnapshot RetentionSnapshot()
    {
        lock (_sync)
        {
            var records = _state.Invocations.Values
                .Select(record => (record.State, HasReceipt: record.Receipt is not null))
                .Concat(_state.Workflows.Values.Select(record => (record.State, HasReceipt: record.Receipt is not null)))
                .ToArray();
            return new CapabilityLedgerRetentionSnapshot(
                records.Count(record => record.State == RecordState.Completed && record.HasReceipt),
                records.Count(record => record.State == RecordState.Completed && !record.HasReceipt),
                records.Count(record => record.State == RecordState.Pending));
        }
    }

    private void Persist()
    {
        try
        {
            var data = JsonSerializer.SerializeToUtf8Bytes(_state, Options);
            var temporary = _filePath + ".tmp";
            File.WriteAllBytes(temporary, data);
            File.Move(temporary, _filePath, overwrite: true);
        }
        catch
        {
            throw Unavailable();
        }
    }

    private static CapabilityBrokerException Unavailable() =>
        new("CAPABILITY_LEDGER_UNAVAILABLE", "ledger");
}

internal enum CapabilityApprovalDecision
{
    Approve,
    Reject
}

internal sealed class CapabilityApprovalStore(TimeSpan? timeToLive = null)
{
    private sealed record IssuedGrant(
        string PlanId,
        string PlanDigest,
        CapabilityPrincipal Principal,
        CapabilityAppContext? AppContext,
        IReadOnlySet<string> Permissions,
        DateTimeOffset ExpiresAt);

    private readonly TimeSpan _timeToLive = timeToLive ?? TimeSpan.FromMinutes(5);
    private readonly Dictionary<string, CapabilityApprovalRequest> _pending = new(StringComparer.Ordinal);
    private readonly Dictionary<string, IssuedGrant> _grants = new(StringComparer.Ordinal);
    private readonly HashSet<string> _consumedTokens = new(StringComparer.Ordinal);
    private readonly object _sync = new();

    public CapabilityApprovalRequest? PendingRequest(string requestId)
    {
        lock (_sync)
        {
            return _pending.GetValueOrDefault(requestId);
        }
    }

    public void DiscardPending(string requestId)
    {
        lock (_sync)
        {
            _pending.Remove(requestId);
        }
    }

    public CapabilityApprovalRequest? Request(
        CapabilityExecutionPlan plan,
        string digest,
        IReadOnlyList<PocketCapabilityDescriptor> descriptors,
        DateTimeOffset now)
    {
        lock (_sync)
        {
            var effects = plan.Steps.Zip(descriptors)
                .Where(pair => pair.Second.ApprovalPolicy.RequiresExecutionApproval())
                .Select(pair => new CapabilityApprovalEffect(
                    pair.First.Id,
                    pair.First.Capability,
                    pair.Second.Effect,
                    CapabilityCanonicalJson.ArgumentsDigest(pair.First.Arguments),
                    $"approval.{pair.First.Capability.Id}",
                    pair.Second.RollbackAvailable))
                .ToArray();
            if (effects.Length == 0)
            {
                return null;
            }
            var permissions = descriptors
                .Where(descriptor => descriptor.ApprovalPolicy.RequiresExecutionApproval())
                .SelectMany(descriptor => descriptor.Permissions)
                .ToHashSet(StringComparer.Ordinal);
            var request = new CapabilityApprovalRequest(
                $"approval:{Guid.NewGuid():D}",
                plan.Id,
                digest,
                plan.Principal,
                plan.AppContext,
                now,
                now.Add(_timeToLive),
                $"nonce:{Guid.NewGuid():D}",
                effects,
                permissions);
            _pending[request.Id] = request;
            return request;
        }
    }

    public CapabilityApprovalGrant Decide(
        string requestId,
        string presentedPlanDigest,
        CapabilityApprovalDecision decision,
        DateTimeOffset now)
    {
        lock (_sync)
        {
            if (!_pending.Remove(requestId, out var request))
            {
                throw new CapabilityBrokerException("CAPABILITY_APPROVAL_INVALID", requestId);
            }
            if (request.PlanDigest != presentedPlanDigest)
            {
                throw new CapabilityBrokerException("CAPABILITY_APPROVAL_INVALID", requestId);
            }
            if (now > request.ExpiresAt)
            {
                throw new CapabilityBrokerException("CAPABILITY_APPROVAL_EXPIRED", requestId);
            }
            if (decision != CapabilityApprovalDecision.Approve)
            {
                throw new CapabilityBrokerException("CAPABILITY_APPROVAL_REJECTED", requestId);
            }
            var token = $"grant:{Guid.NewGuid():D}";
            _grants[token] = new IssuedGrant(
                request.PlanId,
                request.PlanDigest,
                request.Principal,
                request.AppContext,
                request.RequiredPermissions,
                request.ExpiresAt);
            return new CapabilityApprovalGrant(token);
        }
    }

    public void Consume(
        CapabilityApprovalGrant grant,
        CapabilityExecutionPlan plan,
        string digest,
        IReadOnlySet<string> requiredPermissions,
        DateTimeOffset now)
    {
        lock (_sync)
        {
            if (_consumedTokens.Contains(grant.Token))
            {
                throw new CapabilityBrokerException("CAPABILITY_APPROVAL_REPLAYED", grant.Token);
            }
            if (!_grants.Remove(grant.Token, out var issued)
                || issued.PlanId != plan.Id
                || issued.PlanDigest != digest
                || issued.Principal != plan.Principal
                || issued.AppContext != plan.AppContext
                || !requiredPermissions.All(issued.Permissions.Contains))
            {
                throw new CapabilityBrokerException("CAPABILITY_APPROVAL_INVALID", grant.Token);
            }
            _consumedTokens.Add(grant.Token);
            if (_consumedTokens.Count > 1_024)
            {
                _consumedTokens.Clear();
                _consumedTokens.Add(grant.Token);
            }
            if (now > issued.ExpiresAt)
            {
                throw new CapabilityBrokerException("CAPABILITY_APPROVAL_EXPIRED", grant.Token);
            }
        }
    }
}

internal sealed record CapabilityAuditEntry(
    [property: JsonPropertyName("approvalDecision")] string ApprovalDecision,
    [property: JsonPropertyName("approvalPolicy")] string ApprovalPolicy,
    [property: JsonPropertyName("auditEntryId")] string AuditEntryId,
    [property: JsonPropertyName("capability")] CapabilityAuditCapability Capability,
    [property: JsonPropertyName("durationMs")] long DurationMilliseconds,
    [property: JsonPropertyName("idempotencyReplay")] bool IdempotencyReplay,
    [property: JsonPropertyName("inputDigest")] string InputDigest,
    [property: JsonPropertyName("invocationId")] string InvocationId,
    [property: JsonPropertyName("origin")] string Origin,
    [property: JsonPropertyName("permissionDecision")] string PermissionDecision,
    [property: JsonPropertyName("pocketApp")] CapabilityAuditPocketApp? PocketApp,
    [property: JsonPropertyName("principalPseudonym")] string PrincipalPseudonym,
    [property: JsonPropertyName("readback")] CapabilityAuditReadback Readback,
    [property: JsonPropertyName("retryCount")] int RetryCount,
    [property: JsonPropertyName("safeErrorCode")] string? SafeErrorCode,
    [property: JsonPropertyName("status")] string Status,
    [property: JsonPropertyName("timestamp")] DateTimeOffset Timestamp,
    [property: JsonPropertyName("traceId")] string TraceId);

internal sealed record CapabilityAuditCapability(string Id, int Version, string Effect);
internal sealed record CapabilityAuditPocketApp(string Id, string Version, string ManifestDigest);
internal sealed record CapabilityAuditReadback(string Status, string? EvidenceDigest);

internal sealed record CapabilityAuthorizationAuditEntry(
    [property: JsonPropertyName("decision")] string Decision,
    [property: JsonPropertyName("eventType")] string EventType,
    [property: JsonPropertyName("origin")] string Origin,
    [property: JsonPropertyName("planDigest")] string PlanDigest,
    [property: JsonPropertyName("planId")] string PlanId,
    [property: JsonPropertyName("pocketApp")] CapabilityAuditPocketApp? PocketApp,
    [property: JsonPropertyName("principalPseudonym")] string PrincipalPseudonym,
    [property: JsonPropertyName("safeErrorCode")] string? SafeErrorCode,
    [property: JsonPropertyName("timestamp")] DateTimeOffset Timestamp);

internal sealed class CapabilityBrokerAuditLog
{
    private sealed record AuditFile(string Path, DateOnly Day, long ByteCount);

    private static readonly Regex AuditFilePattern = new(
        "^capability-(?<day>[0-9]{8})\\.jsonl$",
        RegexOptions.CultureInvariant | RegexOptions.Compiled);
    private static readonly JsonSerializerOptions Options = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        DefaultIgnoreCondition = JsonIgnoreCondition.Never
    };

    private readonly string _directory;
    private readonly object _sync = new();

    public CapabilityBrokerAuditLog(string rootDirectory)
    {
        try
        {
            _directory = Path.Combine(rootDirectory, "audit");
            Directory.CreateDirectory(_directory);
            if ((File.GetAttributes(_directory) & FileAttributes.ReparsePoint) != 0)
            {
                throw new CapabilityBrokerException("CAPABILITY_AUDIT_UNAVAILABLE", "audit");
            }
        }
        catch
        {
            throw new CapabilityBrokerException("CAPABILITY_AUDIT_UNAVAILABLE", "audit");
        }
    }

    public void Append(CapabilityAuditEntry entry)
    {
        AppendEntry(entry, entry.Timestamp);
    }

    public void AppendAuthorization(CapabilityAuthorizationAuditEntry entry)
    {
        AppendEntry(entry, entry.Timestamp);
    }

    private void AppendEntry<T>(T entry, DateTimeOffset timestamp)
    {
        lock (_sync)
        {
            try
            {
                var path = Path.Combine(_directory, $"capability-{timestamp.UtcDateTime:yyyyMMdd}.jsonl");
                try
                {
                    var attributes = File.GetAttributes(path);
                    if ((attributes & (FileAttributes.Directory | FileAttributes.ReparsePoint)) != 0)
                    {
                        throw new CapabilityBrokerException("CAPABILITY_AUDIT_UNAVAILABLE", "audit");
                    }
                }
                catch (FileNotFoundException)
                {
                }
                catch (DirectoryNotFoundException)
                {
                }
                var data = JsonSerializer.SerializeToUtf8Bytes(entry, Options);
                using var stream = new FileStream(path, FileMode.Append, FileAccess.Write, FileShare.Read);
                stream.Write(data);
                stream.WriteByte((byte)'\n');
                stream.Flush(flushToDisk: true);
            }
            catch
            {
                throw new CapabilityBrokerException("CAPABILITY_AUDIT_UNAVAILABLE", "audit");
            }
        }
    }

    public byte[] CombinedData()
    {
        lock (_sync)
        {
            using var result = new MemoryStream();
            foreach (var file in StrictAuditFiles())
            {
                var data = File.ReadAllBytes(file.Path);
                result.Write(data);
            }
            return result.ToArray();
        }
    }

    public int RemoveEntries(DateTimeOffset? olderThan)
    {
        if (olderThan is null)
        {
            return 0;
        }
        lock (_sync)
        {
            try
            {
                var cutoffDay = DateOnly.FromDateTime(olderThan.Value.UtcDateTime);
                var targets = StrictAuditFiles().Where(file => file.Day < cutoffDay).ToArray();
                foreach (var target in targets)
                {
                    File.Delete(target.Path);
                }
                return targets.Length;
            }
            catch (CapabilityBrokerException)
            {
                throw;
            }
            catch
            {
                throw new CapabilityBrokerException("CAPABILITY_AUDIT_UNAVAILABLE", "audit");
            }
        }
    }

    public int RemoveAllEntries()
    {
        lock (_sync)
        {
            try
            {
                var targets = StrictAuditFiles();
                foreach (var target in targets)
                {
                    File.Delete(target.Path);
                }
                return targets.Count;
            }
            catch (CapabilityBrokerException)
            {
                throw;
            }
            catch
            {
                throw new CapabilityBrokerException("CAPABILITY_AUDIT_UNAVAILABLE", "audit");
            }
        }
    }

    public CapabilityAuditRetentionSnapshot RetentionSnapshot()
    {
        lock (_sync)
        {
            var files = StrictAuditFiles();
            return new CapabilityAuditRetentionSnapshot(files.Count, files.Sum(file => file.ByteCount));
        }
    }

    public static string PrincipalPseudonym(CapabilityPrincipal principal)
    {
        var source = string.Join("\u001f", new[]
        {
            principal.UserId,
            principal.PocketAppId ?? string.Empty,
            principal.AgentSessionId ?? string.Empty
        });
        return "principal:sha256:" + Convert.ToHexString(SHA256.HashData(Encoding.UTF8.GetBytes(source))).ToLowerInvariant();
    }

    private List<AuditFile> StrictAuditFiles()
    {
        try
        {
            var files = new List<AuditFile>();
            foreach (var path in Directory.EnumerateFileSystemEntries(_directory))
            {
                var name = Path.GetFileName(path);
                if (!name.StartsWith("capability-", StringComparison.Ordinal)
                    && !name.EndsWith(".jsonl", StringComparison.Ordinal))
                {
                    continue;
                }
                var match = AuditFilePattern.Match(name);
                if (!match.Success
                    || !DateOnly.TryParseExact(
                        match.Groups["day"].Value,
                        "yyyyMMdd",
                        System.Globalization.CultureInfo.InvariantCulture,
                        System.Globalization.DateTimeStyles.None,
                        out var day))
                {
                    throw new CapabilityBrokerException("CAPABILITY_AUDIT_UNAVAILABLE", "audit");
                }
                var attributes = File.GetAttributes(path);
                if ((attributes & (FileAttributes.Directory | FileAttributes.ReparsePoint)) != 0)
                {
                    throw new CapabilityBrokerException("CAPABILITY_AUDIT_UNAVAILABLE", "audit");
                }
                files.Add(new AuditFile(path, day, new FileInfo(path).Length));
            }
            return files.OrderBy(file => file.Path, StringComparer.Ordinal).ToList();
        }
        catch (CapabilityBrokerException)
        {
            throw;
        }
        catch
        {
            throw new CapabilityBrokerException("CAPABILITY_AUDIT_UNAVAILABLE", "audit");
        }
    }
}
